share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+## `CalleeIsResponsibleForStackCleanup` (`bool`)
+
+Specifies who is responsible for cleaning the stack after the function call. If equal to `true`, it's the callee, otherwise it the caller.
+
+## `FloatsUseGPRs` (`bool`)
+
+Setting this to true disables usage of vector registers entirely.
+
+## `StackArgumentsUseRegularStructAlignmentRules` (`bool`)
+
+Setting this to true allows stack arguments to use slots smaller than a word (makes them behave like a single struct instead of a bunch of arguments): so they don't get extended unnecessarily. A notable ABI using this is apple flavor of the aarch64.
+
+For example
+```cpp
+void function(/* enough arguments of all the registers */, uint8_t, uint8_t);
+```
+When this is set to true, both `uint8_t` will get placed as if they were a struct, that is without any padding in-between. When this is false, both of them will get extended to register size before getting placed, leading to padding.
+
+## `UseStrictAggregateAlignmentRules` (`bool`)
+
+Treat all the structs passed using the stack as if they were aligned unnaturally[^1].
+
+This means that they are going to handled with extreme care (which usually just means extra padding).
+
+## `StackAlignment` (`unsigned`)
+
+States the required alignment of the stack at the point of a function call in bytes.
+
+> **_Note:_** states minimum value for ABIs supporting multiple different stack alignment values, for example, if the ABI requires the stack to be aligned on 4 bytes for internal calls but on 8 bytes for interfaces (like 32-bit ARM ABI), the value of `StackAlignment` should be equal to 4.
+
+
+## `MinimumStackArgumentSize` (`unsigned`)
+
+States the minimum possible stack argument size in bytes. When the actual type of an argument is below this value, some padding with be inserted.
+
+## `StackBytesAllocatedForRegisterArguments` (`unsigned`)
+
+States the number of bytes reserved for the callee to be able to "mirror" register arguments on the stack. This value should be set to `0` for any ABI that does not require such space.
+
+In our context, this serves as the starting offset for the first stack argument. As in, it specifies the number of byte on the top of the stack struct that should be ignored.
+
+## `MaximumGPRsPerAggregateArgument` (`unsigned`)
+
+States the maximum number of GPRs available to pass a single aggregate (a struct, a union, etc.) argument, meaning that it can only be passed in the GPRs if `MaximumGPRsPerAggregateArgument` is less than or equal to the number of the registers required to fit the object including padding.
+
+> **_Note:_** If `MaximumGPRsPerAggregateArgument` is equal to 0, it means that the ABI does not allow aggregate arguments to use GPRs.
+
+> **_Note:_** If an argument doesn't fit into the specified registers or uses irregular padding, the registers are not used and the object is passed using the memory (stack, pointer-to-copy, etc.).
```

Comment:
1. > irregular padding

   What is "irregular padding"? Is it the same as an "unnaturally aligned" `struct`?
2. Is there an ABI where this is not either "all argument registers or none"? That is, could this be a `bool`?
   Same question for the two next options.

Reply:
1. Yeah, as far as I remember, this comment predates our support of unnatural alignment, hence the weird terminology. Thanks for noticing!

2. aarch64 has `2`. There, if it's under 2 words - pass it as is, otherwise, use pointer-to-copy.

Reply:
1. Actually, on a fresh look. We should probably just drop the unnatural alignment part of the note to begin with. Since, unlike when it was written, we now have a dedicated option for that.

Reply:
Or even the entire note? 🤔

It sounds so basic that I'm not sure if it is worth keeping.

Reply:
OK

Reply:
Gone.